### PR TITLE
Support read(Bytebuffer buf) method in LocalCacheFileInStream

### PR DESCRIPTION
**Pull request by neuyilan**
1. HDFS supports the NIO  read (bytebyffer buffer) function, and we'd better be compatible with HDFS API.
2. The `libhdfs.c `client  uses the read (bytebyffer buffer) method to access the HDFS data[1], "**By using Java DirectByteBuffers we can avoid copying the bytes onto the Java heap.  Instead the data will be directly copied from kernel space to the C heap"**. 
[1] https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/hdfs.c#L50

### CODE REVIEW

**Comment by LuQQiu** [comment]
@neuyilan Can you add a description of why this change is needed?

**Comment by neuyilan** [comment]
> @neuyilan Can you add a description of why this change is needed?

I have added the description above, PTAL

File: core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * Implementers of this interface provide a read API that writes to a ByteBuffer, not a byte[].
+ */
+public interface ByteBufferReadable {
+
+  /**
+   * Reads up to buf.remaining() bytes into buf. Callers should use buf.limit(..) to control the
+   * size of the desired read.
+   * <p/>
+   * After a successful call, buf.position() and buf.limit() should be unchanged, and therefore any
+   * data can be immediately read from buf. buf.mark() may be cleared or updated.
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java:26]: this part is different from Hadoop `ByteBufferReadable` (i.e., `After a successful call, {@code buf.position()} will be advanced by the  number of bytes read and {@code buf.limit()} will be unchanged.`) , any reason for that?

File: core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+   * In the case of an exception, the values of buf.position() and buf.limit() are undefined, and
+   * callers should be prepared to recover from this eventuality.
+   * <p/>
+   *
+   * @param buf the ByteBuffer to receive the results of the read operation. Up to buf.limit() -
+   *            buf.position() bytes may be read.
+   * @return the number of bytes available to read from buf
+   * @throws IOException if there is some error performing the read
+   */
+  int read(ByteBuffer buf) throws IOException;
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java:37]: you can leave a default implementation here, like

```suggestion
default int read(ByteBuffer buf) throws IOException {
  throw new UnsupportedOperationException("read(ByteBuffer buf) not implemented");
}
```

Then you don't need to implement individuals in other classes

File: core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ */
+package alluxio.client;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+/**
+ * Implementers of this interface provide a read API that writes to a ByteBuffer, not a byte[].
+ */
+public interface ByteBufferReadable {
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java:19]: Check `alluxio.client.file.FileInStream` where we already define and provided the default implementation of `  public int read(ByteBuffer byteBuffer, int off, int len) throws IOException`.

You can either move this method there (and remove interface `ByteBufferReadable`), or move that methd here. I don't have a strong opinion. But I do think they should be put together

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java [hunk 4]
```diff
@@ -346,4 +347,80 @@ private static void registerGauges() {
...
+          mCacheManager.put(pageId, page, mCacheScope, mCacheQuota);
+        }
+      }
+    }
+    if (totalBytesRead > len || (totalBytesRead < len && remaining() > 0)) {
+      throw new IOException(String.format("Invalid number of bytes read - "
+              + "bytes to read = %d, actual bytes read = %d, bytes remains in file %d",
+          len, totalBytesRead, remaining()));
+    }
+    buf.put(b, off, len);
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java:423]: it looks like `buf.position()` will be changed after a successful call, right?

**Review by apc999 — CHANGES_REQUESTED** [review]
Left some questions. I think this can be a very useful optimization and let's iterate on this.

Thanks!

File: core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java (hunk 1 above)
Comment (neuyilan) [reply core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java]: Maybe it's written by mistake. In fact, when the call is successful, `buf.position()` will change, I have altered the comment

File: core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java (hunk 2 above)
Comment (neuyilan) [reply core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java]: Sure, Thanks.

File: core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java (hunk 3 above)
Comment (neuyilan) [reply core/client/fs/src/main/java/alluxio/client/ByteBufferReadable.java]: Yeah, as there already one `public int read(ByteBuffer byteBuffer, int off, int len) throws IOException` method in `alluxio.client.file.FileInStream`, so I think it's ok to move the interface's logic to `alluxio.client.file.FileInStream` class.

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java (hunk 4 above)
Comment (neuyilan) [reply core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java:381]: Yes :)

**Comment by neuyilan** [comment]
> Left some questions. I think this can be a very useful optimization and let's iterate on this.
> 
> Thanks!

Thank you for your review. I have fixed some of your comments, sorted out the redundant code, and added some UTs, PTAL.

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java [hunk 5]
```diff
@@ -287,19 +265,41 @@ private FileInStream getExternalFileInStream(long pos) throws IOException {
...
+        page = readExternalPageIntoByteBuffer(pos);
+        break;
+      default:
+        throw new IOException("unsupported read type = " + readType);
+    }
+    return page;
+  }
+
+  /**
+   * Using read(byte b[], int off, int len) method to read data from external storage which contains
```
Comment (LuQQiu) [inline core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java:296]: Using -> Uses

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java [hunk 6]
```diff
@@ -320,6 +320,46 @@ private FileInStream getExternalFileInStream(long pos) throws IOException {
     return page;
   }
 
+  /**
+   * Using read(ByteBuffer buf) method to read data from external storage which contains the
```
Comment (LuQQiu) [inline core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java:324]: Using -> Uses

File: core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java [hunk 7]
```diff
@@ -346,4 +386,16 @@ private static void registerGauges() {
           });
     }
   }
+
```
Comment (LuQQiu) [inline core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java:389]: Remove extra line

**Review by LuQQiu — COMMENTED** [review]
left some minor codestyle suggestions

**Review by neuyilan — COMMENTED** [review]
> left some minor codestyle suggestions

Thanks @LuQQiu, I found that we have our code style file in `build/checkstyle/alluxio_checks.xml`, but those may still not check the indentations format problems like [1], and every code developer may have its own code styles, if all the checks need to be protected by the person who reviews the code, there will inevitably be loopholes. So do we considering using automatic code checking and code collation like spotless[2]?
It's just a discussion, any comments are welcome!

[1] https://github.com/Alluxio/alluxio/pull/13406/files/61f6121d395af64bcc1088bcd6cbbeaef33164ad#diff-ab09d305553e29cf84278bdb023cb71470918975d4ac0f725084e38bb0c8a317
[2]https://github.com/diffplug/spotless

**Review by calvinjia — COMMENTED** [review]
Would it be possible to check for the preferred API and use that if available instead of having a switch based on readtype?

**Review by calvinjia — APPROVED** [review]
Discussed offline - this PR is just to add the interface so applications can start integrating, optimization of the interface with the streams underneath (ie. external stream) will happen later.

**Review by apc999 — APPROVED** [review]

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Pull request merged**